### PR TITLE
Add UrlFromDoiProcessor

### LINKED ISSUE #37: implement UrlFromDoiProcessor
```bib
@article{
  doi = {xyz}
}
```

```php
use RenanBr\BibTexParser\Processor\UrlFromDoiProcessor;
$listener->addProcessor(new UrlFromDoiProcessor());
```

Expected result:

```
Array
(
    [0] => Array
        (
            [doi] => xyz
            [url] => https://doi.org/xyz
        )
)
```

### PR DESCRIPTION
**Pull request by LeJeanbono**
Fix #37

### CODE REVIEW

File: src/Processor/UrlFromDoiProcessor.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+     * @param array $entry
+     *
+     * @return array
+     */
+    public function __invoke(array $entry)
+    {
+        $covered = $this->getCoveredTags(array_keys($entry));
+        foreach ($covered as $tag) {
+            if ($tag === 'doi') {
+                $entry['url'] = 'https://doi.org/' . $entry[$tag];
```
Comment (renanbr) [inline src/Processor/UrlFromDoiProcessor.php]: - What do you think if we make the `url` value configurable via constructor? and keep defaults to `url`
- I would prefer it to move `https://doi.org/` to a class constant

**Comment by renanbr** [comment]
What do you think it should be the behaviour for these scenarios?

```php
['doi' => 'xyz', 'url' => 'http://example.org']; // prex-existing url
['doi' => ''];                                   // empty doi
```

a) exception
b) `url` to null
c) do nothing
d) something else?

I tend to say `c`, WDYT?

File: src/Processor/UrlFromDoiProcessor.php [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace RenanBr\BibTexParser\Processor;
+
+class UrlFromDoiProcessor
+{
+    use TagCoverageTrait;
```
[the remark this answers is not in the document]
Comment (renanbr) [reply src/Processor/UrlFromDoiProcessor.php]: entry tags are case insensitive, use `TagSearchTrait`, then we can call `tagSearch()` to get the real tag name for `doi` tag :wink:

File: tests/Processor/UrlFromDoiProcessorTest.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace RenanBr\BibTexParser\Test\Processor;
+
+use PHPUnit\Framework\TestCase;
+use RenanBr\BibTexParser\Processor\UrlFromDoiProcessor;
+
+class UrlFromDoiProcessorTest extends TestCase
```
Comment (renanbr) [inline tests/Processor/UrlFromDoiProcessorTest.php]: Could you add tests for these cases that processor should bypass?
```
['doi' => 'xyz', 'url' => 'http://example.org']; // prex-existing url
['doi' => ''];                                   // empty doi
```

**Comment by LeJeanbono** [comment]
> What do you think it should be the behaviour for these scenarios?
> 
> ```
> ['doi' => 'xyz', 'url' => 'http://example.org']; // prex-existing url
> ['doi' => ''];                                   // empty doi
> ```
> 
> a) exception
> b) `url` to null
> c) do nothing
> d) something else?
> 
> I tend to say `c`, WDYT?

I don't know bibtex but with some research, if `url` and `doi` are already present, they should be has :
`url = "https://doi.org/10.1038/421805a"` and `doi = "10.1038/421805a"`
So => `c`

**Comment by LeJeanbono** [comment]
Empty `doi` => Exception ?

File: src/Processor/UrlFromDoiProcessor.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+use RenanBr\BibTexParser\Exception\ProcessorException;
+
+class UrlFromDoiProcessor
+{
+    use TagSearchTrait;
+
+    private $urlDoiPrefix;
+
+    public function __construct($urlDoiPrefix = 'https://doi.org')
```
Comment (renanbr) [inline src/Processor/UrlFromDoiProcessor.php]: That's very nice, I liked the `$urlDoiPrefix` argument

Could you make some changes ?

- Place default value in a class constant, as `https://doi.org/%s`
- Change default argument value to `null`, then pick it from the constant when necessary
- Change its name (and property) to `$urlDoiFormat`, then use `sprintf()` in `__invoke()`

File: src/Processor/UrlFromDoiProcessor.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+     */
+    public function __invoke(array $entry)
+    {
+        $doiTag = $this->tagSearch('doi', array_keys($entry));
+        $urlTag = $this->tagSearch('url', array_keys($entry));
+        if ($urlTag === null && $doiTag !== null) {
+            $doiValue = $entry[$doiTag];
+            if ($doiValue === '') {
+                throw new ProcessorException('doi tag should not be empty');
+            }
```
Comment (renanbr) [inline src/Processor/UrlFromDoiProcessor.php]: let's just skip it :wink:

Comment (LeJeanbono) [reply src/Processor/UrlFromDoiProcessor.php]: Ok

File: src/Processor/UrlFromDoiProcessor.php (hunk 4 above)
Comment (LeJeanbono) [reply src/Processor/UrlFromDoiProcessor.php]: Done !

**Comment by renanbr** [comment]
thank you @LeJeanbono !
merged as https://github.com/renanbr/bibtex-parser/commit/1704405568f353046c456de3ce11c360a1eb2723 and https://github.com/renanbr/bibtex-parser/commit/173bbf134ec639d7e5e762ed549c2e5482f138a4

**Comment by LeJeanbono** [comment]
You're welcome !

**Pull request closed without merging**